### PR TITLE
[SPARK-36369][PYTHON] Fix Index.union to follow pandas 1.3

### PR DESCRIPTION
**Pull request by itholic**
### What changes were proposed in this pull request?

This PR proposes fixing the `Index.union` to follow the behavior of pandas 1.3.

Before:
```python
>>> ps_idx1 = ps.Index([1, 1, 1, 1, 1, 2, 2])
>>> ps_idx2 = ps.Index([1, 1, 2, 2, 2, 2, 2])
>>> ps_idx1.union(ps_idx2)
Int64Index([1, 1, 1, 1, 1, 2, 2], dtype='int64')
```

After:
```python
>>> ps_idx1 = ps.Index([1, 1, 1, 1, 1, 2, 2])
>>> ps_idx2 = ps.Index([1, 1, 2, 2, 2, 2, 2])
>>> ps_idx1.union(ps_idx2)
Int64Index([1, 1, 1, 1, 1, 2, 2, 2, 2, 2], dtype='int64')
```

This bug is fixed in https://github.com/pandas-dev/pandas/issues/36289.

### Why are the changes needed?

We should follow the behavior of pandas as much as possible.

### Does this PR introduce _any_ user-facing change?

Yes, the result for some cases have duplicates values will change.

### How was this patch tested?

Unit test.

### CODE REVIEW

**Comment by itholic** [comment]
cc @HyukjinKwon @ueshin @xinrong-databricks

File: python/pyspark/pandas/indexes/base.py [hunk 1]
```diff
@@ -2235,6 +2235,24 @@ def union(
         """
         Form the union of two Index objects.
 
+        .. note:: For duplicated values, pandas chooses the number of duplicates of self or other
+            with more duplicates. But counting all duplicates is very expensive for large data,
+            so pandas-on-Spark always chooses the number of duplicates in self.
```
Comment (HyukjinKwon) [inline python/pyspark/pandas/indexes/base.py:2240]: Is there any release note in pandas?

File: python/pyspark/pandas/indexes/base.py [hunk 2]
```diff
@@ -2293,8 +2311,6 @@ def union(
         sdf_self = self._internal.spark_frame.select(self._internal.index_spark_columns)
         sdf_other = other_idx._internal.spark_frame.select(other_idx._internal.index_spark_columns)
         sdf = sdf_self.union(sdf_other.subtract(sdf_self))
```
Comment (ueshin) [inline python/pyspark/pandas/indexes/base.py:2313]: Could you try `sdf_self.union(sdf_other).exceptAll(sdf_self.intersectAll(sdf_other))`?

It seems working at least with the example above:

```py
>>> idx1 = ps.Index([1, 1, 2, 2, 2])
>>> idx2 = ps.Index([1, 1, 1, 2, 2])
>>>
>>> idx1.union(idx2)
Int64Index([1, 1, 1, 2, 2, 2], dtype='int64')
>>> idx1.to_pandas().union(idx2.to_pandas())
Int64Index([1, 1, 1, 2, 2, 2], dtype='int64')
```

I haven't tried the other examples and I'm not sure how expensive this is, though.

File: python/pyspark/pandas/indexes/base.py (hunk 1 above)
Comment (itholic) [reply python/pyspark/pandas/indexes/base.py:2240]: They mentioned it's bug fix in their release note at https://pandas.pydata.org/pandas-docs/dev/whatsnew/v1.3.0.html#indexing.

Let me try the solution @ueshin commented https://github.com/apache/spark/pull/33634#discussion_r682859453

File: python/pyspark/pandas/indexes/base.py (hunk 2 above)
Comment (itholic) [reply python/pyspark/pandas/indexes/base.py:2313]: Let me try. Thanks!

**Review by xinrong-meng — APPROVED** [review]

**Review by ueshin — APPROVED** [review]

File: python/pyspark/pandas/tests/indexes/test_base.py [hunk 3]
```diff
@@ -1542,6 +1540,19 @@ def test_union(self):
                     pidx2.union(pd.Series([1, 2, 3, 4, 3, 4, 3, 4])),
                     almost=True,
```
Comment (HyukjinKwon) [inline python/pyspark/pandas/tests/indexes/test_base.py:1541]: Wouldn't the test now fail with pandas lower than 1.3? We should at least skip or make the tests pass manually in other versions

File: python/pyspark/pandas/indexes/base.py [hunk 4]
```diff
@@ -2292,9 +2292,7 @@ def union(
 
         sdf_self = self._internal.spark_frame.select(self._internal.index_spark_columns)
         sdf_other = other_idx._internal.spark_frame.select(other_idx._internal.index_spark_columns)
-        sdf = sdf_self.union(sdf_other.subtract(sdf_self))
-        if isinstance(self, MultiIndex):
-            sdf = sdf.drop_duplicates()
+        sdf = sdf_self.unionAll(sdf_other).exceptAll(sdf_self.intersectAll(sdf_other))
         if sort:
```
Comment (HyukjinKwon) [inline python/pyspark/pandas/indexes/base.py:2296]: @itholic, did you refer to this?

> Bug in Index.union() and MultiIndex.union() dropping duplicate Index values when Index was not monotonic or sort was set to False (GH36289, GH31326, GH40862)

It sounds like now it drops duplicates but the codes above look keeping the duplicates. Should we also at least check `sort` and drop duplicates?

File: python/pyspark/pandas/tests/indexes/test_base.py [hunk 5]
```diff
@@ -1527,10 +1527,8 @@ def test_union(self):
                 almost=True,
             )
 
+            # Bug in Index.union dropping duplicated values is fixed in pandas 1.3
             if LooseVersion(pd.__version__) >= LooseVersion("1.3"):
```
Comment (HyukjinKwon) [inline python/pyspark/pandas/tests/indexes/test_base.py:1531]: can we remove this and share one test base?

**Review by HyukjinKwon — COMMENTED** [review]
otherwise it looks good to me too

File: python/pyspark/pandas/indexes/base.py (hunk 4 above)
Comment (itholic) [reply python/pyspark/pandas/indexes/base.py:2296]: Thanks, @HyukjinKwon !

Seems like they said "dropping duplicate Index values" was a bug, and fixed in this release.

Related discussion was made at https://github.com/pandas-dev/pandas/issues/36289.

Also checked the `sort` and drop duplicates, works the same as pandas.

File: python/pyspark/pandas/tests/indexes/test_base.py (hunk 5 above)
Comment (itholic) [reply python/pyspark/pandas/tests/indexes/test_base.py]: Thanks!!

**Review by HyukjinKwon — APPROVED** [review]

File: python/pyspark/pandas/tests/indexes/test_base.py (hunk 5 above)
Comment (ueshin) [reply python/pyspark/pandas/tests/indexes/test_base.py]: Hmm, I'm afraid I'm against the change 6ed03e7.
It doesn't test with the actual pandas behavior we are following, so even if pandas changes the behavior in the future, we won't know it.

Comment (HyukjinKwon) [reply python/pyspark/pandas/tests/indexes/test_base.py]: Can we remove the old test cases, and only keep the tests only with pandas 1.3?

Comment (HyukjinKwon) [reply python/pyspark/pandas/tests/indexes/test_base.py]: We can leverage `unittest.skipIf`. I would prefer to avoid having duplicated test cases.

Comment (itholic) [reply python/pyspark/pandas/tests/indexes/test_base.py]: Oh, then do we want to use

```python
    @unittest.skipIf(
        LooseVersion(pd.__version__) < LooseVersion("1.3")
    )
```

and remove all the version related logics such as `if LooseVersion(pd.__version__) >= LooseVersion("0.24"):` and `if LooseVersion(pd.__version__) >= LooseVersion("1.0.0")` ??

Comment (HyukjinKwon) [reply python/pyspark/pandas/tests/indexes/test_base.py]: Hm .. okay maybe it's a separate to topic to work on/discuss. I am fine with reverting the specific change.

Comment (itholic) [reply python/pyspark/pandas/tests/indexes/test_base.py]: Sounds good. Just reverted to the previous fix.

**Comment by HyukjinKwon** [comment]
Merged to master and branch-3.2.

**Pull request closed without merging**